src/core/execute.c

```diff
@@ -210,6 +210,64 @@ bool exec_needs_ipc_namespace(const ExecContext *context) {
         return context->private_ipc || context->ipc_namespace_path;
 }
 
+static bool can_apply_cgroup_namespace(const ExecContext *context, const ExecParameters *params) {
+        int r;
+
+        r = cg_all_unified();
+        if (r < 0) {
+                log_once_exec_warning_errno(context, params, r, "Failed to determine cgroup hierarchy version: %m");
+                return false;
+        }
+        if (r == 0) {
+                log_once_exec_warning(context, params, "ProtectControlGroups=%s is configured, but the unified cgroups hierarchy is not set up, ignoring namespace setup.",
+                                      protect_control_groups_to_string(context->protect_control_groups));
+                return false;
+        }
+
+        if (!ns_type_supported(NAMESPACE_CGROUP)) {
+                log_once_exec_warning(context, params, "ProtectControlGroups=%s is configured, but the kernel does not support cgroup namespaces, ignoring namespace setup.",
+                                      protect_control_groups_to_string(context->protect_control_groups));
+                return false;
+        }
+        return true;
+}
+
+static inline bool needs_cgroup_namespace(ProtectControlGroups i) {
+        return IN_SET(i, PROTECT_CONTROL_GROUPS_PRIVATE, PROTECT_CONTROL_GROUPS_STRICT);
+}
+
+ProtectControlGroups get_adjusted_protect_control_groups(const ExecContext *context, const ExecParameters *params) {
```

Comment:
I think `_adjusted_` is not necessary, but please prefix `exec_`. So, `exec_get_protect_control_groups()` should be ok.